heat/core/statistics.py

```diff
@@ -626,77 +627,94 @@ def maximum(x1, x2, out=None):
     >>> ht.maximum(a, d)
     ValueError: operands could not be broadcast, input shapes (3, 4) (3, 4, 5)
     """
-    # perform sanitation
-    if not isinstance(x1, dndarray.DNDarray) or not isinstance(x2, dndarray.DNDarray):
-        raise TypeError(
-            "expected x1 and x2 to be a ht.DNDarray, but were {}, {} ".format(type(x1), type(x2))
-        )
-    if out is not None and not isinstance(out, dndarray.DNDarray):
-        raise TypeError("expected out to be None or an ht.DNDarray, but was {}".format(type(out)))
-
-    # apply split semantics
-    if x1.split is not None or x2.split is not None:
-        if x1.split is None:
-            x1.resplit_(x2.split)
-        if x2.split is None:
-            x2.resplit_(x1.split)
-        if x1.split != x2.split:
-            if np.prod(x1.gshape) < np.prod(x2.gshape):
-                x1.resplit_(x2.split)
-            if np.prod(x2.gshape) < np.prod(x1.gshape):
-                x2.resplit_(x1.split)
+    # TODO: move all of this to __binary_op
+    # sanitize input
+    in_x = [x1, x2]
+    splits = []
+    shapes = []
+    comms = []
+    devices = []
+    t_x = []
+    for i, x in enumerate(in_x):
+        if isinstance(x, (int, float)):
+            x = factories.array([x], dtype=types.canonical_heat_type(type(x)), split=None)
+            in_x[i] = x
+        if not isinstance(x, dndarray.DNDarray):
+            raise TypeError("expected two DNDarrays, but array {} was {} ".format(i, type(x)))
+        if x.size == 0:
+            raise ValueError("operation not possible, array {} contains no data".format(i))
+        splits.append(x.split)
+        shapes.append(x.gshape)
+        comms.append(x.comm)
+        devices.append(x.device.device_type)
+        t_x.append(x._DNDarray__array)
+
+    # sanitize devices
+    if devices.count("gpu") == 1:
+        not_on_gpu = abs(devices.index("gpu") - 1)
+        # move to gpu
+        in_x[not_on_gpu] = in_x[not_on_gpu].gpu()
+
+    # sanitize shapes
+    output_gshape = stride_tricks.broadcast_shape(shapes[0], shapes[1])
+
+    # sanitize splits
+    if not splits.count(None) == 2:
+        # DNDarrays must be distributed along same axis
+        if splits.count(None) == 0:
+            if splits.count(splits[0]) == len(splits):
```

Comment:
i really like this logic here, but do we need to count in the last if loop? isnt this the same as `splits[0] == splits[1]` since its only of length two? this is just to make it slightly easier to read, i dont think that there will be any noticable speedup here so feel free to ignore this if you like it the way it is

Reply:
You're absolutely right, no need to be so generic here, I'll change it.